tools/mpremote/mpremote/mip.py

```diff
@@ -62,6 +64,11 @@ def _rewrite_url(url, branch=None):
 
 
 def _download_file(transport, url, dest):
+    if not url.startswith(allowed_mip_url_prefixes):
+        print("Installing:", dest, "from", url)
+        _ensure_path_exists(transport, dest)
+        transport.fs_put(url, dest, progress_callback=show_progress_bar)
```

Comment:
I just realised `fs_put` was removed in db59e55fe7a0b67d3af868990468e7b8056afe42

It should now use `fs_writefile` (signature is different though).

I suggest refactoring this function to the following:
```py
if url.startswith(allowed_mip_url_prefixes):
    try:
        with urllib.request.urlopen(url) as src:
            data = src.read()
    except urllib.error. ...
        # existing urllib error handling
else:
    with open(url, "rb") as f:
        data = f.read()

print("Installing:", dest)
_ensure_path_exists(transport, dest)
transport.fs_writefile(dest, data, progress_callback=show_progress_bar)
```

Reply:
Resolved.

Reply:
I have also added some code for correct handling of files installed from the filesystem on Windows.

Reply:
Was delayed for a few days as I did not have access to my windows machine for testing.